### PR TITLE
Runtime interfaces

### PR DESCRIPTION
**Pull request by hidalgopl**
Podman features (updated: 11/30/2020):

- [x] /rest/v1/updatepod
- [x] /rest/v1/logs - without follow
- [x] /rest/v1/status

Outstading tasks:
- [x] Run podman e2e tests as part of CI/CD [link](https://github.com/elotl/itzo/pull/107/files#diff-7146b45fc8ae58c143328f2c98de48eccac0873c143712fb848371bcf9ddcd3c)
- [ ] Move logic inside `server.Server.runAttach` to `ItzoRuntime.Attach` (this method isn't part of interface yet) 
- [ ] Implement port-forwarding with podman runtime
- [ ] Implement `logs --follow` with podman runtime
- [ ] Implement exec with podman runtime 
- [ ] Implement attach with podman runtime

### CODE REVIEW

File: pkg/runtime/runtime.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
+package runtime
+
+import (
+	"github.com/elotl/itzo/pkg/api"
+	"github.com/elotl/itzo/pkg/logbuf"
+)
+
```
Comment (hidalgopl) [inline pkg/runtime/runtime.go:7]: @ldx @jrroman pinging you guys here, as this PR isn't ready yet, but those interfaces shouldn't change much. Let me know what you think about the design.

I think next step will be refactoring itzo runtime (I've already started that in this PR) and making sure that behavior is unchanged and tests are passing. Once it's done I can finish podman's implementation of those interfaces (should be pretty straight forward). Any feedback much appreciated;)

File: pkg/util/image.go [hunk 2]
```diff
@@ -37,3 +38,29 @@ func ParseImageSpec(image string) (string, string, error) {
 	}
 	return server, repo, err
 }
+
```
Comment (hidalgopl) [inline pkg/util/image.go:41]: everything below in this PR (from `pkg/util/`) are just util functions that were living in `pod_controller.go`; I moved them to util package as well as tests to slim down `pod_controller.go`

File: pkg/convert/convert.go [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+		DNSPolicy:        v1.DNSPolicy(podSpec.DNSPolicy),
+		SecurityContext:  &podSecurityContext,
+		ImagePullSecrets: nil,
+		Hostname:         podSpec.Hostname,
+		Subdomain:        podSpec.Subdomain,
+		HostAliases:      hostAliases,
+		DNSConfig:        &podDNSConfig,
+	}
+	return spec
+}
```
[the remark this answers is not in the document]
Comment (hidalgopl) [reply pkg/convert/convert.go]: Yeah, I cleaned up `convert.go` here today, so only needed things are there. Copied also conversion functions from there with small modifications.

File: pkg/convert/convert.go [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+					Host: probe.Handler.TCPSocket.Host,
+				},
+			},
+			InitialDelaySeconds: probe.InitialDelaySeconds,
+			TimeoutSeconds:      probe.TimeoutSeconds,
+			PeriodSeconds:       probe.PeriodSeconds,
+			SuccessThreshold:    probe.SuccessThreshold,
+			FailureThreshold:    probe.FailureThreshold,
+		}
+	}
```
[the remark this answers is not in the document]
Comment (hidalgopl) [reply pkg/convert/convert.go:147]: Copied one from kip :+1: thanks for pointing this one out.

File: pkg/convert/convert.go [hunk 5]
```diff
@@ -0,0 +1,330 @@
+package convert
+
+import (
+	"github.com/containers/libpod/v2/libpod/define"
+	"github.com/elotl/itzo/pkg/api"
+	"k8s.io/api/core/v1"
+	v12 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
[the remark this answers is not in the document]
Comment (hidalgopl) [reply pkg/convert/convert.go]: Yeah, missed this one, for some reason my IDE picked that alias. Fixed

File: pkg/runtime/podman.go [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+	connText context.Context
+}
+
+func (p *PodmanImageService) ListImages() {
+	return
+}
+
+func (p *PodmanImageService) ImageStatus(rootdir, image string) error {
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (hidalgopl) [reply pkg/runtime/podman.go:104]: I actually deleted those in downstream PR: https://github.com/elotl/itzo/pull/107

File: pkg/runtime/podman.go [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+var (
+	restartPolicyMap = map[api.RestartPolicy]string{
+		api.RestartPolicyAlways: RestartPolicyAlways,
+		api.RestartPolicyOnFailure: RestartPolicyOnFailure,
+		api.RestartPolicyNever: RestartPolicyNo,
+	}
+)
+
+type PodmanSandbox struct {
+	connText context.Context
```
[the remark this answers is not in the document]
Comment (hidalgopl) [reply pkg/runtime/podman.go:42]: Yes - that naming is taken from [podman go bindings tutorial](https://github.com/containers/podman/blob/master/docs/tutorials/podman-go-bindings.md)

File: pkg/convert/convert.go [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+	vol := v1.Volume{
+		Name: volume.Name,
+		VolumeSource: v1.VolumeSource{
+			HostPath: &v1.HostPathVolumeSource{
+				Path: path,
+				Type: &hostPathType,
+			},
+		},
+	}
+	return vol
```
[the remark this answers is not in the document]
Comment (hidalgopl) [reply pkg/convert/convert.go]: Copied volume conversion from KIP. I also added function that converts PackagePath source to HostPath

**Comment by hidalgopl** [comment]
> This kinda resembles the CRI, but we ditched all the API types and only kept the function names. This adds the complexity of the large number of possible calls, but not the benefits of actual CRI compatibility.
I think we should with either a simple interface where there's less of a chance of missing something and less boilerplate added, or the actual CRI of that brings benefits.

I actually simplified this a bit in #107 , that's my latest take on those inferfaces: https://github.com/elotl/itzo/blob/95dc176568f71b26e18dffc029a7c758f6a9ace2/pkg/runtime/runtime.go
Let me know if you think this is simple enough, otherwise I can iterate on this till we get it to the point we all like the internal API design ;)

**Review by ldx — APPROVED** [review]
Cool! Let's merge this, then focus on:
- E2E tests for itzo
- Moving exec, attach and portforward to the runtime package.

**Pull request merged**